functions/src/modules/gmail.functions.ts

```diff
@@ -0,0 +1,273 @@
+          labelIds: [doneLabelObj.id],
+          topicName: 'projects/audio-seva-team-test/topics/gmail-labeled-done',
+        },
+      });
+
+      if (watchResults.data && doneLabelObj) {
+        // Store the most recent historyId
+        await storeHistoryIdInDatabase(
+          emailAddress,
+          watchResults.data.historyId
+        );
+        return res
+          .status(200)
+          .send(
+            `Watch (${watchResults.data.historyId}) created on ${
+              doneLabelObj.name
+            }/${doneLabelObj.id} for ${emailAddress}`
+          );
+      } else {
+        return res
+          .status(301)
+          .send(`Something went wrong, check logs for Gmail-initWatch`);
+      }
+    } catch (err) {
+      console.log('Error: ', err);
+      return res
+        .status(301)
+        .send(`Something went wrong, check logs for Gmail-initWatch`);
+    }
+  }
+);
+
+export const gmailDoneHandler = functions.pubsub
+  .topic('gmail-labeled-done')
+  .onPublish(async (message: Message) => {
+    const decodedMessage = JSON.parse(
+      Buffer.from(message.data, 'base64').toString('ascii')
+    );
+
+    // Gmail client and authenticate
+    const { client_key, secret, redirect } = functions.config().gmail;
+    const oauth2Client = new Google.google.auth.OAuth2(
+      client_key,
+      secret,
+      redirect
+    );
+    const { oauth, lastSyncHistoryId } = await fetchToken(
+      decodedMessage.emailAddress
+    );
+    oauth2Client.setCredentials({
+      access_token: oauth.token,
+      refresh_token: oauth.refreshToken,
+    });
+    const gm = await Google.google.gmail({
+      version: 'v1',
+      auth: oauth2Client,
+    });
+
+    const labelResults = await gm.users.labels.list({
+      userId: 'me',
+    });
+    const doneLabelObj = labelResults.data.labels.filter(label => {
+      return label.name === 'SQRDone';
+    })[0];
+
+    try {
+      // Typescript complains if not initialized and casted to any
+      const historyListRequest: any = {
+        userId: decodedMessage.emailAddress || 'me',
+        startHistoryId: lastSyncHistoryId,
+        historyTypes: 'labelAdded',
+        labelId: [doneLabelObj.id],
+      };
+
+      // Call the history list api to fetch changes since last synced
+      const results = await gm.users.history.list(historyListRequest);
+      if (!results.data.history || !results.data.history.length) {
+        // Handle empty history, and update history id
+        await storeHistoryIdInDatabase(
+          decodedMessage.emailAddress,
+          decodedMessage.historyId
+        );
+        throw new Error('History is empty. Nothing to process');
+      }
+
+      const onlyLabelAdded = results.data.history.filter(change => {
+        return change.labelsAdded;
+      });
+
+      // We only care about the final state of the Done label whether it was applied
+      // or not, so we get rid of the duplicates incase there are any
+      const uniques = {};
```

Comment:
Isn't it possible to get uniques by `reduce`?

Reply:
Yes, everything is possible using a reduce. Though it would still be 2 loops right?

Reply:
OK, I'm not pushing. Keep it as it is if you feel it's OK.